### ISSUE TITLE
Scale parameter downcasted and rounded down in pytorch.distributions.Normal

### ISSUE THREAD

**Issue by markocai** [pytorch/pytorch#38474]
## 🐛 Bug

Normal distribution's scale parameter is casted from float to int and rounded down. This is a bug or at least very unexpected, undocumented behavior. Documentation of the `pytorch.distributions.Normal` does not mention casting `loc` and `scale` to the same data type and  potentially causing rounding errors. 

## To Reproduce

1. Create a normal distribution with the following parameters: `loc` is `Tensor` with `dtype=int`, `scale` is `float`.
2. Values of the distribution parameters are casted to integers and `scale` parameter is rounded down. 

Example code 
```
import torch
import torch.distributions as td
td.Normal(torch.tensor(1), 0.5)
```

returns
```
Normal(loc: 1, scale: 0)
```

## Expected behavior
Normal distribution with the given values of parameters and the data types of the parameters casted to the float.

In the example `Normal(loc: 1.0, scale: 0.5)` is expected.

Another way to formulate this is that behavior for the following three alternatives should be same

```
(Normal(torch.tensor(1), torch.tensor(0.5)), 
Normal(torch.tensor(1), 0.5), 
Normal(1, torch.tensor(0.5))

(Normal(loc: 1, scale: 0.5),
 Normal(loc: 1, scale: 0),
 Normal(loc: 1.0, scale: 0.5))
```

## Environment

 - PyTorch Version (e.g., 1.0):  1.4.0
 - OS (e.g., Linux): Mac OSX 10.15.4
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.7
 - CUDA/cuDNN version: No CUDA

## Additional context
This can lead to difficult to track down bugs and probably affects also other probability distribution implementations in PyTorch.

**Comment by cpaulse** [comment]
The integer cast is happening inside `broadcast_all()`, which is applying the type of the first encountered tensor to any subsequent `numbers.Number` instances. So, any other distribution that processes its input this way (`Beta`, `Binomial`, `Uniform`, ...)   will show the same behavior.

**Comment by markocai** [comment]
> The integer cast is happening inside `broadcast_all()`, which is applying the type of the first encountered tensor to any subsequent `numbers.Number` instances. So, any other distribution that processes its input this way (`Beta`, `Binomial`, `Uniform`, ...) will show the same behavior.

True. This is the way how the input is processed in `pytorch.distributions.Normal`. Would it be possible to fix this behavior?

**Comment by cpaulse** [comment]
> > The integer cast is happening inside `broadcast_all()`, which is applying the type of the first encountered tensor to any subsequent `numbers.Number` instances. So, any other distribution that processes its input this way (`Beta`, `Binomial`, `Uniform`, ...) will show the same behavior.
> 
> True. This is the way how the input is processed in `pytorch.distributions.Normal`. Would it be possible to fix this behavior?

I'm looking into a fix.

**Comment by fritzo** [comment]
I think it's probably best to handle this before `broadcast_all()` in case some distributions need a long tensor. Maybe something like
```py
def as_float(x):
    if isinstance(x, torch.Tensor):
        if x.dtype in (torch.long, torch.int, torch.bool):  # Is there an alias?
            x = x.to(dtype=torch.get_default_dtype())  # Is there a less gross way?
    elif isinstance(x, int):
        x = float(x)
    return x

class Normal(Distribution):
    def __init__(self, loc, scale, ...):
        loc = as_float(loc)
        scale = as_float(scale)
        loc, scale = broadcast_all(loc, scale)
        ...
```
What helpers are there to correctly coerce int-like types to float-like types of proper precision?

**Comment by cpaulse** [comment]
> I think it's probably best to handle this before `broadcast_all()` in case some distributions need a long tensor. Maybe something like
> 
> ```python
> def as_float(x):
>     if isinstance(x, torch.Tensor):
>         if x.dtype in (torch.long, torch.int8, torch.bool):  # Is there an alias?
>             x = x.to(dtype=torch.get_default_dtype())  # Is there a less gross way?
>     elif isinstance(x, int):
>         x = float(x)
>     return x
> 
> class Normal(Distribution):
>     def __init__(self, loc, scale, ...):
>         loc = as_float(loc)
>         scale = as_float(scale)
>         loc, scale = broadcast_all(loc, scale)
>         ...
> ```
> 
> What helpers are there to correctly coerce int-like types to float-like types of proper precision?

You could use `torch.is_floating_point(x)` in place of `x.dtype in (torch.long, torch.int8, torch.bool)` I guess, but yes I like the idea of not messing with `broadcast_all()`.
